frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+
+	/// Number of eras to check per block.
+	///
+	/// If set to 0, this pallet does absolutely nothing.
+	///
+	/// Based on the amount of weight available at `on_idle`, up to this many eras of a single
+	/// nominator might be checked.
+	#[pallet::storage]
+	pub type ErasToCheckPerBlock<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// The events of this pallet.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A staker was unstaked.
+		Unstaked { stash: T::AccountId, maybe_pool_id: Option<PoolId>, result: DispatchResult },
+		/// A staker was slashed for requesting fast-unstake whilst being exposed.
+		Slashed { stash: T::AccountId, amount: BalanceOf<T> },
+		/// A staker was partially checked for the given eras, but the process did not finish.
+		Checked { stash: T::AccountId, eras: Vec<EraIndex> },
+		/// Some internal error happened while migrating stash. They are removed as head as a
+		/// consequence.
+		Errored { stash: T::AccountId },
+		/// An internal error happened. Operations will be paused now.
+		InternalError,
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq))]
+	pub enum Error<T> {
+		/// The provided Controller account was not found.
+		///
+		/// This means that the given account is not bonded.
+		NotController,
+		/// The bonded account has already been queued.
+		AlreadyQueued,
+		/// The bonded account has active unlocking chunks.
+		NotFullyBonded,
+		/// The provided un-staker is not in the `Queue`.
+		NotQueued,
+		/// The provided un-staker is already in Head, and cannot deregister.
+		AlreadyHead,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_idle(_: T::BlockNumber, remaining_weight: Weight) -> Weight {
+			if remaining_weight.any_lt(T::DbWeight::get().reads(2)) {
+				return Weight::from_ref_time(0)
+			}
+
+			Self::do_on_idle(remaining_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register oneself for fast-unstake.
+		///
+		/// The dispatch origin of this call must be signed by the controller account, similar to
+		/// `staking::unbond`.
+		///
+		/// The stash associated with the origin must have no ongoing unlocking chunks. If
+		/// successful, this will fully unbond and chill the stash. Then, it will enqueue the stash
+		/// to be checked in further blocks.
+		///
+		/// If by the time this is called, the stash is actually eligible for fast-unstake, then
+		/// they are guaranteed to remain eligible, because the call will chill them as well.
+		///
+		/// If the check works, the entire staking data is removed, i.e. the stash is fully
+		/// unstaked, and they potentially join a pool with their entire bonded stake.
+		///
+		/// If the check fails, the stash remains chilled and waiting for being unbonded as in with
+		/// the normal staking system, but they lose part of their unbonding chunks due to consuming
```

Comment:
So this basically means that if the check succeeds - we don't charge for the extra resources consumed? Could that be a potential hole for spamming?
But then again you'd always know who is doing it and then they risk their funds.

Reply:
> So this basically means that if the check succeeds

Yes, if the checks succeeds, essentially it is free. this could be abused, but scope is very limited, since at most, this pallet will consume resources on_idle. Meaning that if the chain actually has other transactions wanting to fill the block, this pallet will not do anything (even if `ErasToCheckPerBlock == infinity`)